File: dmoj/vmss/vmss7wc16c3p2/vmss7wc16c3p2.ml
open Base

let buildGraph n m =
  let graph = Array.create ~len:(n + 1) [] in
  let () =
    for _ = 1 to m do
      let (u, v) = Stdlib.Scanf.scanf "%i %i\n" (fun u v -> (u, v)) in
      Array.set graph u (v :: graph.(u));
      Array.set graph v (u :: graph.(v))
    done
  in
  graph

let rec dfs graph u goal visited =
  (not visited.(u))
  && (phys_equal u goal
     ||
     let () = Array.set visited u true in
     List.fold_until ~init:false
       ~f:(fun accumulator v ->
         if accumulator then Stop true else Continue (dfs graph v goal visited))
       ~finish:Fn.id graph.(u))

let () =
  let (n, m, a, b) =
    Stdlib.Scanf.scanf "%i %i %i %i\n" (fun n m a b -> (n, m, a, b))
  in
  let graph = buildGraph n m in
  let visited = Array.create ~len:(n + 1) false in
  if dfs graph a b visited then Stdio.printf "GO SHAHIR!\n"
  else Stdio.printf "NO SHAHIR!\n"

File: dmoj/vmss/vmss7wc16c3p2/vmss7wc16c3p2.cpp
#include <iostream>
#include <vector>

using namespace std;

const int MAX_N = 2000;
vector<int> graph[MAX_N + 1];
bool visited[MAX_N + 1] = {false};

bool dfs(int u, int goal) {
    if (visited[u]) {
        return false;
    }
    if (u == goal) {
        return true;
    }

    visited[u] = true;
    bool ans = false;
    for (int v : graph[u]) {
        ans |= dfs(v, goal);
    }

    return ans;
}

int main() {
    int n, m, a, b;
    scanf("%i %i %i %i", &n, &m, &a, &b);

    for (int i = 0; i < m; i++) {
        int u, v;
        scanf("%i %i", &u, &v);
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    if (dfs(a, b)) {
        printf("GO SHAHIR!\n");
    } else {
        printf("NO SHAHIR!\n");
    }
}
